File: backend-go/module/show/model/show.go
package showmodel

import (
	"cinema/common"
	"time"
)

const EntityName = "Show"
const TableName = "shows"

type Show struct {
	ID           int                `json:"id" gorm:"column:id;primary_key" swaggerignore:"true"`
	Date         *common.Date       `json:"date" gorm:"column:date" `
	StartTime    *common.Time       `json:"startTime" gorm:"column:start_time"`
	EndTime      *common.Time       `json:"endTime" gorm:"column:end_time"`
	AuditoriumID int64              `json:"-" gorm:"column:auditorium_id"`
	Auditorium   *common.Auditorium `json:"auditorium" gorm:"preload:false;foreignKey:AuditoriumID;references:ID;constraint:OnUpdate:CASCADE,OnDelete:SET NULL;"`
	ImdbID       string             `json:"imdbID" gorm:"column:imdb_id"`
	CreatedAt    *time.Time         `json:"-" gorm:"column:created_at"`
	UpdatedAt    *time.Time         `json:"-" gorm:"column:updated_at"`
}

func (Show) TableName() string { return TableName }

func (s *Show) Mask(isAdminOrOwner bool) {
	if auditorium := s.Auditorium; auditorium != nil {
		auditorium.Mask(isAdminOrOwner)
	}
}

type ShowCreate struct {
	ID               int          `json:"id" gorm:"column:id;primary_key" swaggerignore:"true"`
	Date             *common.Date `json:"date" gorm:"column:date" `
	StartTime        *common.Time `json:"startTime" gorm:"column:start_time"`
	EndTime          *common.Time `json:"endTime" gorm:"column:end_time"`
	AuditoriumID     int64        `json:"-" gorm:"column:auditorium_id"`
	AuditoriumFakeID string       `json:"auditoriumID" gorm:"preload:false;foreignKey:AuditoriumID;references:ID;constraint:OnUpdate:CASCADE,OnDelete:SET NULL;"`
	ImdbID           string       `json:"imdbID" gorm:"column:imdb_id"`
}

func (ShowCreate) TableName() string { return TableName }

type ShowUpdate struct {
	Date         *common.Date `gorm:"column:date" json:"date"`
	EndTime      *common.Time `gorm:"column:end_time" json:"endTime"`
	StartTime    *common.Time `gorm:"column:start_time" json:"startTime"`
	AuditoriumID int64        `gorm:"column:auditorium_id" json:"-"`
	ImdbID       string       `gorm:"column:imdb_id" json:"imdbID"`
}

func (ShowUpdate) TableName() string { return TableName }

File: frontend/src/models/show.ts
import Auditorium from './auditorium';
import Cinema from "./cinema";
import Movie from './movie';

export default interface Show {
  id: string;
  auditorium: Auditorium;
  date: string;
  endTime: string;
  imdbID: string;
  startTime: string;
  movie: Movie;
}

export interface AuditoriumShow extends Auditorium {
  shows: Show[];
}
export interface CinemaAuditorium extends Cinema {
  auditoriums: AuditoriumShow[];
}
export class MovieShow {
  private cinemas: Map<string, CinemaAuditorium> = new Map();

  constructor(shows: Show[]) {
    for (const show of shows) {
      const cinemaId = show.auditorium.cinema.id;
      const auditoriumId = show.auditorium.id;

      // Create or get CinemaAuditorium from map
      let cinemaAuditorium = this.cinemas.get(cinemaId);
      if (!cinemaAuditorium) {
        cinemaAuditorium = {
          ...show.auditorium.cinema,
          auditoriums: [],
        };
        this.cinemas.set(cinemaId, cinemaAuditorium);
      }

      // Find or create AuditoriumShow
      let auditoriumShow = cinemaAuditorium.auditoriums.find(
        (aud) => aud.id === auditoriumId,
      );
      if (!auditoriumShow) {
        auditoriumShow = {
          ...show.auditorium,
          shows: [],
        };
        cinemaAuditorium.auditoriums.push(auditoriumShow);
      }

      // Add show to AuditoriumShow
      auditoriumShow.shows.push(show);
    }
  }

  getCinemas(date?: string): CinemaAuditorium[] {
    if (date) {
      return [...this.cinemas.values()]
        .map((cinema) => {
          cinema.auditoriums = cinema.auditoriums
            .map((auditorium) => {
              auditorium.shows = auditorium.shows.filter(
                (show) => show.date === date,
              );
              return auditorium;
            })
            .filter((auditorium) => auditorium.shows.length > 0);
          return cinema;
        })
        .filter((cinema) => cinema.auditoriums.length > 0);
    }
    return [...this.cinemas.values()];
  }
}
